content/blog/package-react-components/index.md

```diff
@@ -0,0 +1,776 @@
+
+Unfortunately in early 2020 there still no universal standard way of Packaging a Front End module with all
+its assets, but a lot of progress has been made and there are actually lots of promising solutions.
+
+We are going to explore how this problem is tackled in the wild by studying
+each technique and providing example real world production ready libraries that
+use them and in the end of this post we are going
+to recommend a particular solution to this problem based of my experience.
+
+Additionally we are going to explore how each of these packaging schemes enable
+different **Theming** schemes and, at the end, recommend one.
+
+Most of this post is also applicable to any piece of Front End Javascript code that is related
+to CSS, images, etc, such as [Angular][3] Components / Directives, [Vue][15] Components, etc.
+
+## To Inline or not to Inline
+
+In the context of Front End assets, inlining means to bake those assets directly
+into the Javascript code (bundle) generated by our bundlers or compilers such as Webpack or [Babel][4].
+
+CSS can be inlined without any third party library but to enable all regular CSS features
+a third party **CSS-in-JS** library is recommended such as [Emotion][5] or [Styled-components][6], more on this later.
+Images can be inlined by Webpack's [url-loader][11] in the form of [Data URL][100] with base 64 encoding.
+SVG can be inlined by turning an `.svg` file into a [React][1] Component (An SVG file is just a regular HTML tree)
+or by using Webpack's [inline-svg-loader][101]
+
+The main advantage of inlining is that after you bundle all your code base you end up with
+a single Javascript file that contains all the dependencies it needs to run, even CSS and Images.
+
+This makes _very_ easy to package and publish to [NPM][102] and it makes very easy
+for **Component Consumers** to _use_ your module.
+
+Using **CSS-in-JS** libraries like [Emotion][5] or [Styled-components][6] makes up for a very
+nice **Run-Time** **Theming** API. Since everything is Javascript your Themes are regular Javascript
+object that can be composed or broken down into parts as any other regular JS object,
+and these can be altered at Run-Time, Themes can be composed, and they provide
+a much simpler strategy than Build-Time CSS generated Themes like [Bootstrap][9] or [Ant Design][8].
+[Material UI React][7] uses a Run-Time Theme via a custom **CSS-in-JS** solution.
+
+Inlining is particularly useful for Component libraries or even UI Frameworks because
+they tend to be mostly CSS heavy and with a couple of SVGs at most, both of which are efficiently
+inlined.
+
+Images are a bit more problematic because they are inlined in [base 64][103] making them 30% bigger.
+
+In my experience, most Component Libraries do not tend to use an over abundance
+of images and if they do they tend to be rather small.
+
+Most high resolution images are dynamic content that are usually dealt by
```

Comment:
remil nit: I'm pretty sure that because _content_ is an uncountable noun, the person is singular (_most high resolution images are dynamic content that **is** usually_)

Reply:
fixed!